### PR TITLE
Update Similarity Transformer Architecture to compute both embeddings

### PR DESCRIPTION
**Pull request by activatedgeek**
This PR takes steps towards a better architecture by incorporating a conditional graph operation to allow switching between the code embedding network and the string embedding network.

---
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/examples/233)

### CODE REVIEW

**Comment by jlewi** [comment]
@activatedgeek What's the status of this PR?

Is there any way to right a unittest to make sure that this works as expected? 

What if we mock out/override encode? Would that allow us to replace a transformer encoder with something known (maybe just pass the features through?)? So we would know what the expected output is.

**Comment by activatedgeek** [comment]
I've actually been using this branch to test out my variations and particularly to get a solution out for #259.

File: code_search/src/code_search/t2t/similarity_transformer.py [hunk 1]
```diff
@@ -73,5 +72,9 @@ def encode(self, features, input_key):
 
   def infer(self, features=None, **kwargs):
     del kwargs
+
+    if 'targets' not in features:
```
Comment (jlewi) [inline code_search/src/code_search/t2t/similarity_transformer.py]: Why do we need to check if targets is in features? 
Is the prediction in inference mode the concatenation of code and query embeddings?
If it is then do we need this if statement because the user can control what embeddings are computed
1) setting inputs and targets to zeros or non-zero values 
2) Which elements of the output vector they look at.

File: code_search/src/code_search/t2t/query.py [hunk 2]
```diff
@@ -14,16 +14,19 @@ def get_encoder(problem_name, data_dir):
   return problem.feature_info["inputs"].encoder
 
 
-def encode_query(encoder, query_str):
+def encode_query(encoder, query_str, embed_code=False):
```
Comment (jlewi) [inline code_search/src/code_search/t2t/query.py]: Can we get rid of embed_code now that we are concatenating the two vectors and computing both embeddings?

**Comment by jlewi** [comment]
@activatedgeek It looks like this PR is making two changes

1. Change the loss function to fix #259 
1. Fix model export to allow #260 to compute embeddings

I believe #2 will necessitate changes to other parts of the code because the output of inference will now be a vector
[query_embeddings, code_embeddings]

So would it make sense to get an initial version of this PR merged so we can begin making the other changes to the code?

**Comment by jlewi** [comment]
@activatedgeek And thank you so much for continuing to work on this! Great to see this coming along.

File: code_search/src/code_search/t2t/similarity_transformer.py (hunk 1 above)
Comment (activatedgeek) [reply code_search/src/code_search/t2t/similarity_transformer.py]: The eval step during the training also uses this function and deliberately does not send in targets. I added this to make sure they work. Certainly not ideal though.

File: code_search/src/code_search/t2t/query.py (hunk 2 above)
Comment (activatedgeek) [reply code_search/src/code_search/t2t/query.py]: This actually leads to a bunch of changes here and even the Dataflow step before this. Shall we defer this change to the final PR after we make sure the model works?

**Comment by activatedgeek** [comment]
@jlewi I think we can go ahead and merge this.

I'm happy to be working on this, albeit slower than I'd like.

File: code_search/src/code_search/t2t/similarity_transformer.py (hunk 1 above)
Comment (activatedgeek) [reply code_search/src/code_search/t2t/similarity_transformer.py]: It is kind of the client's responsibility to slice and dice the I/O vectors. This part is mostly safeguard the eval step of T2T.

Comment (jlewi) [reply code_search/src/code_search/t2t/similarity_transformer.py]: Can you put that in a comment?

File: code_search/src/code_search/t2t/query.py (hunk 2 above)
Comment (jlewi) [reply code_search/src/code_search/t2t/query.py]: Defer which change the whole PR or just removing embed_code?

**Comment by jlewi** [comment]
Regarding your earlier comment

> This actually leads to a bunch of changes here and even the Dataflow step before this. Shall we defer this change to the final PR after we make sure the model works?

Was that referring to getting rid of embed_config or this PR as a whole?

I assume once we merge this PR a lot of the Dataflow code will break and need to be updated?

Any suggestions about how we can go about fixing the code?

Would it be possible to start writing unittests to verify the various Dataflow transforms are working with the new model? Perhaps we could support that by adding a test utility function to emit dummy models for the new model architecture?

File: code_search/src/code_search/t2t/query.py [hunk 3]
```diff
@@ -14,16 +14,19 @@ def get_encoder(problem_name, data_dir):
...
   even though it is ignored during inference.
   See tensorflow/tensor2tensor#868"""
 
   encoded_str = encoder.encode(query_str) + [text_encoder.EOS_ID]
-  features = {"inputs": tf.train.Feature(int64_list=tf.train.Int64List(value=encoded_str)),
-              "targets": tf.train.Feature(int64_list=tf.train.Int64List(value=[0]))}
+  features = {
+      "inputs": tf.train.Feature(int64_list=tf.train.Int64List(value=encoded_str)),
+      "targets": tf.train.Feature(int64_list=tf.train.Int64List(value=[0])),
+      "embed_code": tf.constant(embed_code, dtype=tf.bool)
```
Comment (jlewi) [inline code_search/src/code_search/t2t/query.py]: This looks like a bug. I think FeatureProto can only be Int,Bytes, and Floats.
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/example/feature.proto

Should "inputs" be the concatenation of the encoded query and the encoded targets now that we are treating output in prediction as concatenation of the two vectors?

File: code_search/src/code_search/t2t/similarity_transformer.py (hunk 1 above)
Comment (jlewi) [reply code_search/src/code_search/t2t/similarity_transformer.py]: I think this is causing problems during inference using TFServing.

When I send predictions to TFServing I'm getting b'{ "error": "You must feed a value for placeholder tensor \\\'Placeholder\\\' with dtype int64 and shape [?,?,1,1]\\n\\t [[{{node Placeholder}} = Placeholder[_output_shapes=[[?,?,1,1]], dtype=DT_INT64, shape=[?,?,1,1], _device=\\"/job:localhost/replica:0/task:0/device:CPU:0\\"]()]]" }'

**Comment by jlewi** [comment]
I think we can close this in favor of #291.

I originally based #291 on this PR but I think I fixed the issues with the using conditionals.

1. The inputs will always be provided by the "inputs" feature; we will just use different embedding variable scope
1. We need to reduce the feature used as the predicate to a rank 0 vector. I think this may have caused the problems during eval.

**Comment by activatedgeek** [comment]
I think this PR has diverged enough to be closed and all the newer code is more relevant anyways. Let me close this.
